runtime-light/crypto/crypto-builtins.cpp

```diff
@@ -19,10 +38,118 @@ task_t<Optional<string>> f$openssl_random_pseudo_bytes(int64_t length) {
   buffer.store_trivial<uint32_t>(TL_GET_CRYPTOSECURE_PSEUDORANDOM_BYTES);
   buffer.store_trivial<int32_t>(length);
 
+  // TODO think about performance when transferring data
 
-  string request_buf{};
+  string request_buf;
   request_buf.append(buffer.data(), buffer.size());
 
   auto query = f$component_client_send_query(string("crypto"), request_buf);
-  co_return co_await f$component_client_get_result(co_await query);
+  string resp = co_await f$component_client_get_result(co_await query);
+
+  buffer.clean();
+  buffer.store_bytes(resp.c_str(), resp.size());
+
+  std::optional<uint32_t> magic = buffer.fetch_trivial<uint32_t>();
+  if (!magic || *magic != TL_RESULT_TRUE) {
+    co_return false;
+  }
+  std::string_view str_view = buffer.fetch_string();
+  co_return string(str_view.data(), str_view.size());
+}
+
+task_t<Optional<array<mixed>>> f$openssl_x509_parse(const string &data, bool shortnames) {
+  tl::TLBuffer buffer;
+  buffer.store_trivial<uint32_t>(TL_GET_PEM_CERT_INFO);
+  buffer.store_trivial<uint32_t>(shortnames ? TL_BOOL_TRUE : TL_BOOL_FALSE);
+  buffer.store_string(std::string_view(data.c_str(), data.size()));
+
+  string request_buf;
+  request_buf.append(buffer.data(), buffer.size());
+
+  auto query = f$component_client_send_query(string("crypto"), request_buf);
+  string resp_from_platform = co_await f$component_client_get_result(co_await query);
+
+  buffer.clean();
+  buffer.store_bytes(resp_from_platform.c_str(), resp_from_platform.size());
+
+  if (const auto magic = buffer.fetch_trivial<uint32_t>(); magic.value_or(TL_ZERO) != TL_RESULT_TRUE) {
+    co_return false;
+  }
+
+  if (const auto magic = buffer.fetch_trivial<uint32_t>(); magic.value_or(TL_ZERO) != TL_DICTIONARY) {
+    co_return false;
+  }
+
+  const std::optional<uint32_t> size = buffer.fetch_trivial<uint32_t>();
+  if (!size) {
+    co_return false;
+  }
+
+  auto response = array<mixed>::create();
+
+  for (uint32_t i = 0; i < size; ++i) {
+    const auto key_view = buffer.fetch_string();
+    if (key_view.empty()) {
+      co_return false;
+    }
+
+    string key(key_view.data(), key_view.length());
+
+    const std::optional<uint32_t> magic = buffer.fetch_trivial<uint32_t>();
+    if (!magic) {
```

Comment:
Do you check that optional is empty or magic is not equal to zero? Why not just call `has_value` method?

Reply:
Previously I used `has_value`, but it seemed to me that `operator!()` is better. 
I changed auto on optional to emphasize that casting to `bool` is for `std::optional` semantic.
But `has_value()` is better, I agree now